flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala

```diff
@@ -47,9 +47,22 @@ object ScalarOperators {
       nullCheck: Boolean,
       resultType: TypeInformation[_],
       left: GeneratedExpression,
-      right: GeneratedExpression)
-    : GeneratedExpression = {
-    val leftCasting = numericCasting(left.resultType, resultType)
+      right: GeneratedExpression): GeneratedExpression = {
+
+    val leftCasting = operator match {
+      case "%" =>
+        if (left.resultType == right.resultType) {
+          numericCasting(left.resultType, resultType)
+        } else {
+          val castedType = if (isDecimal(left.resultType)) {
+            Types.LONG
```

Comment:
Wouldn't it be better to use `BigDecimal. remainder` instead of losing precision?

Reply:
`BigDecimal. remainder` accept a `BigDecimal` ` divisor` and return a `BigDecimal` result. when `right operand` is INTEGER and `left operand` is DECIMAL，`%` should return INTEGER. So may be using `BigDecimal. remainder` is not suitable. what do you think?

Reply:
Yes sounds reasonable.